### PR TITLE
Refactor database session handling to async with retry logic, remove middleware

### PR DESCRIPTION
**Pull request by ahdamin**
The PR refactors database session management to improve the performance and avoid continuous `PendingRollbackError` raised after the [loss of database connection](https://github.com/Clinical-Genomics/genotype-api/issues/148) `Database connection error: (pymysql.err.OperationalError) (2013, 'Lost connection to MySQL server during query')` 

### Context & design choices
Since `genotype-api` utilizes FastAPI, which is **asynchronous** by design, that makes it a bit sensitive to session sharing issues. Multiple requests in an asynchronous environment can overlap and potentially share the session and may lead to issues such as `PendingRollbackError`.
The **dependency injection** for database sessions was enough for FastAPI. In Flask, middleware with a global session works because each request runs in its own thread, and the session is scoped to that thread.

When using SQLAlchemy's `AsyncSession`, **Lazy Loading** breaks because it does not natively use `await`, and you may encounter `MissingGreenlet`. The eager loading was implemented using the `selectinload` method to avoid the error, without affecting the logic of the service/endpoint.

In some cases, when implementing async queries, an explicit **join condition** was added because the code switched from accessing the object’s relationship to using a query.

### Fixed

 Handles [lost database connection](https://github.com/Clinical-Genomics/genotype-api/issues/148)

### CODE REVIEW

File: genotype_api/security.py [hunk 1]
```diff
@@ -68,8 +68,22 @@ async def get_active_user(
...
+            detail="Invalid or expired token",
+        )
+
+    payload = token_info.get("payload")
+    if not payload or "email" not in payload:
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Invalid or expired token",
+        )
+
```
Comment (ahdamin) [inline genotype_api/security.py:84]: Fixes: TypeError: `NoneType` object is not subscriptable

File: genotype_api/database/crud/read.py [hunk 2]
```diff
@@ -1,114 +1,159 @@
...
+from genotype_api.database.models import SNP, Analysis, Genotype, Plate, Sample, User
 
 LOG = logging.getLogger(__name__)
 
 
 class ReadHandler(BaseHandler):
 
-    def get_analyses_by_plate_id(self, plate_id: int) -> list[Analysis]:
-        analyses: Query = self._get_query(Analysis)
+    async def get_analyses_by_plate_id(self, plate_id: int) -> list[Analysis]:
```
Comment (ChrOertlin) [inline genotype_api/database/crud/read.py:34]: do we need these all un the CRUD layer? shouldn't the api layer be enough?

**Review by ChrOertlin — APPROVED** [review]
THanks for the in depth explanation, I learned from that!

Just curious about the async and await propagation. Do we need that throughout he whole code or is it enough to have it in the api when it calls the top level function of a multi layer?

**Comment by ahdamin** [comment]
> Just curious about the async and await propagation. Do we need that throughout the whole code or is it enough to have it in the api when it calls the top level function of a multi layer call chain?
> 
> I.e. would this be enough?
> 
> ```python
> async def get_something():
>        await do_the_get()
>        return
> 
> ## the do the get
> def do_the_get():
>       return the_gotten
> ```
> 
> instead of also having the async and await in the `do_the_get()`
> 
> Edit: I asked GPT seems to be dependent on the underlying function and whether their operation are synchronous or asynchronous of nature.


I understand your point, and it would be simpler to use only `async` and `await` at the API level. However, as you mentioned, it depends on the nature of each function. If a function performs async operations, then both the function itself and any functions calling it need to be marked **async** and use **await**.

**Comment by Vince-janv** [comment]
> I understand your point, and it would be simpler to use only `async` and `await` at the API level. However, as you mentioned, it depends on the nature of each function. If a function performs async operations, then both the function itself and any functions calling it need to be marked **async** and use **await**.

I'm not convinced adding async/await to each function and return is necessary. It feels weird that to use FastAPI one needs to drastically change the syntax in every python function. Should it not be added only to the endpoints which take a lot of time?

**Comment by ahdamin** [comment]
> I'm not convinced adding async/await to each function and return is necessary. It feels weird that to use FastAPI one needs to drastically change the syntax in every python function. Should it not be added only to the endpoints which take a lot of time?

Of course, it is not necessary to change every Python function. Async operations are related to IO operations such as database operations, file operations, etc. Some functions were not changed here.
Can we go for a synchronous FastAPI application? Of course, we can! The question is why?

File: genotype_api/database/crud/read.py [hunk 3]
```diff
@@ -1,114 +1,159 @@
...
+            selectinload(Analysis.genotypes), selectinload(Analysis.sample)
+        )
         filter_functions = [AnalysisFilter.BY_PLATE_ID]
-        return apply_analysis_filter(
+        filtered_query = apply_analysis_filter(
             analyses=analyses, filter_functions=filter_functions, plate_id=plate_id
-        ).all()
+        )
+        result = await self.session.execute(filtered_query)
+        return result.scalars().all()
```
[the remark this answers is not in the document]
Comment (ahdamin) [reply genotype_api/database/crud/read.py:43]: That is a very valid question. However, it is about how SQLAlchemy handles query execution in an asynchronous environment.
In a **synchronous** environment, SQLAlchemy allows you to chain `.all()` or `.first()` directly on the query to execute it and fetch the results. It blocks the code until the result is available. On the other hand, in **async**, you need to await the query execution with `session.execute(query)`, and then you can retrieve the results using `.scalars().all()`.  Please check this SqlAlchmey [doc](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#querying).

 Please let me know if you find a way to make it look a bit better

**Pull request merged**